Provenance

Nat1405 opened the pull request:
This merge is to prepare Nifty data products for CAOM2 ingestion. This is mostly updating FITS header values, and in the case of reduced cube products, adding `cal` extensions that list the calibrations used.

Open questions:
- Should we archive the files used to produce the telluric correction as well, or the telluric cubes/ extracted telluric spectra?
- Ditto for the flux calibration.

Nat1405 committed:
Changes processed ronchi name to have a _ronchi suffix.

Nat1405 committed:
Changes name of processed arc to include _arc suffix.

Nat1405 committed:
Renames final calibrations to preserve original filename.

Nat1405 committed:
Tags dataproducts with new extensions.

Tags reduced datacubes with a "CAL" extension that identifies calibrations used. Also adds the bad pixel mask (*_bpm.pl) as a fits extension that can be extracted and used as an input bad pixel mask in gemini iraf tasks, and updates the product datalabel.

Nat1405 committed:
Reorders copying of calibrations to support metadata additions.

Nat1405 committed:
Updates checkLists to rewrite text lists. This preps for tagging calibrations.

Nat1405 committed:
Adds tagging of calibrations with header keywords and bpm extension.

Nat1405 committed:
Adds integration tests for nifsBaselineCalibration.

Nat1405 committed:
Updates capitalizations of datalabels.

Fixes #29.

Nat1405 committed:
Adds short nifsBaselineCalibration int test.

This also checks that datalabels and prefixes only contain 'g' or 'G' if the frame is the result of a co-add with n>1 frames in gemcombine.

Nat1405 committed:
Updates 'G' prefix, and fixes caom2 metadata.

Fixes #28 and #29.

Nat1405 committed:
Fixes TravisCI iraf problems.

SharonGoliath commented on nifty/pipeline/steps/nifsBaselineCalibration.py:
```diff
@@ -782,25 +782,27 @@ def makeRonchi(ronchilist, ronchiflat, calflat, grating, over, flatdark, log):
         if over:
             iraf.delete("ronchifile")
-            os.remove("rgn"+ronchiflat+".fits")
+            os.remove("rgn"+ronchiflat+"_ronchi.fits")
```
You asked a developer for comments, so you're going to get developer-centric comments :)

There are seven places in this pull request where the file name is hard-coded. This makes for brittle, hard-to-maintain code. 

There's a principle of software development, known as "Don't Repeat Yourself". What that looks like in this case is: define a variable that contains the file name, and then use the variable everywhere you would otherwise have changed the file name.

Having said that, I don't know what kind of guidance you were given, along with this code, so if you were told not to make un-necessary changes, take this with a grain of salt. Or if you haven't got time left to make these changes, ignore this comment.

SharonGoliath commented on nifty/pipeline/nifsUtils.py:
```diff
@@ -550,22 +550,16 @@ def copyCalibrationDatabase(inputPrefix, grating, fileType, over):
-                        newName = item.split('_')
-                        newName = inputPrefix[:2]+fileType+'_SCI_'+newName[2]+'_'
-                        shutil.copy(item, scienceDirectory+"/calibrations/database/"+newName)
+                        shutil.copy(item, scienceDirectory+"/calibrations/database/")
```
Nice code cleanup.

SharonGoliath commented on nifty/pipeline/nifsUtils.py:
```diff
@@ -607,7 +595,8 @@ def replaceNameDatabaseFiles(inputFile, oldFileName, newFileName):
         f.write(modifiedLines)
-    shutil.move("temp.txt", inputFile)
+    os.remove(inputFile)
+    shutil.move("temp.txt", re.sub(oldFileName, newFileName, inputFile))
```
I'm sure I'm missing something, but the old code changed the content of the file, and the new code changes the file name, at least, based on shutil.move documentation, that's what I think it's doing. Is that what you want?

SharonGoliath commented on nifty/pipeline/steps/nifsReduce.py:
```diff
@@ -220,36 +220,39 @@ def start(kind, telluricDirectoryList="", scienceDirectoryList=""):
+                iraf.delete(os.getcwd()+'/'+os.path.split(arc)[-1]+".fits")
                 # Copy the spatial calibration arc flat frame from Calibrations_grating to the observation directory.
-                shutil.copy(os.getcwd()+'/calibrations/finalArc.fits', arc+'.fits')
+                shutil.copy(os.getcwd()+'/'+arc+'.fits', os.getcwd()+'/'+os.path.split(arc)[-1]+".fits")
```
Is the os.getcwd() call required?

SharonGoliath commented on nifty/pipeline/nifsUtils.py:
```diff
@@ -1338,5 +1338,121 @@ def __init__(self, frame):
+        return any([x.name == "CAL" for x in hdulist])
+
+    def hasBPMExt(self, file):
+        hdulist = astropy.io.fits.open(file)
```
My knowledge of astropy.io.fits is rudimentary, so apologies if this is an obvious question - can the call to 'hasBPMExt` be in the scope of the 'with open' structure right before it, so that the file does not have to be opened twice? Or does the flush call ruin the in-memory access?

If the second open is necessary, should add a close.

SharonGoliath commented on nifty/pipeline/nifsUtils.py:
```diff
@@ -1460,5 +1462,151 @@ def hasBPMExt(self, file):
+        pass
+
+    def hasBPMExt(self, file):
+        hdulist = astropy.io.fits.open(file)
```
Same comment about the additional astropy.io.fits.open call as in the previous commit.

SharonGoliath commented on nifty/pipeline/nifsUtils.py:
```diff
@@ -1460,5 +1462,151 @@ def hasBPMExt(self, file):
+
+        lines = [x.rstrip('\n') + '.fits' for x in lines]
+
+        assert len(lines) > 0
```
'assert' statements in code - I can't find the link right now, but python versions exist where asserts can cause spurious failures. And then you have to decide whether using asserts in production code is a good idea. CADC in general does not, but I don't know the intended production environment for this code.

In any case, as a general rule, provide a comment to the assertion as well, to give the failure case more context, for the future, when you don't remember why you were making the check in the first place.

SharonGoliath commented on nifty/pipeline/nifsUtils.py:
```diff
@@ -1460,5 +1462,151 @@ def hasBPMExt(self, file):
+            calfile = glob.glob(calpath)[0]
+        except IndexError as e:
+            logging.error("Calibration file {} not found.".format(calpath))
+            raise CalibrationsError()
```
Do you intend to replace the well-known FileNotFoundError with a custom CalibrationsError? If the error raised intentionally changes downstream processing, that might be ok, but otherwise you've added un-necessary code and complexity.

I don't know off the top of my head if os.path.exists is a cheaper call than glob.glob, but using glob is an interesting way of finding out if a path exists.

SharonGoliath commented on nifty/pipeline/nifsUtils.py:
```diff
@@ -1516,15 +1522,19 @@ def tagFlat(self, cals):
-                    hdul['PRIMARY'].header['FCMB'+str(i+1)] = (os.path.split(cals.flats[i])[1], self.keywordDict['FCMB'])
+                if len(cals.flats) > 1:
+                    hdul['PRIMARY'].header['FCOMBINE'] = (len(cals.flats), self.keywordDict['FCOMBINE'])
+                    for i in range(len(cals.flats)):
```
Confirming that you mean for the primary header to NOT have ACMB?, FCMB?, DCMB? keywords when there's only a single input?

andamian commented on .travis.yml:
```diff
@@ -28,4 +28,5 @@ install:
 
 script:
-  - pytest
+  - travis_wait 30 pytest nifty/tests/test_nifsSort.py
```
Not sure why the wait is required here. Can the functionality be achieved through other means? Typically one expects `pytest` to run all the tests in the `tests` directory whereas this approach requires explicit specification of test files.

Nat1405 replied:
Hi Adrian, thanks so much for the feedback, sorry it took me so long to rework this.

I believe I've fixed the `wait` problem in test_nifsSort.py. 

The `-s` in the second call to pytest comes up because IRAF/PyRAF tasks still listen to stdin even when run in full automatic mode and error out (I believe) because they try to read from it during pyest. I see two possible solutions:

- Pass an empty txt file as stdin to IRAF tasks like so (might be safer that way to not let accidental stdin get through. Although it would require going through many, many IRAF tasks and making sure stdin is set up properly.):
```python
# Set up/prepare IRAF.
    with open('tmp.txt', 'w') as f:
        pass

    iraf.gemini(Stdin="tmp.txt")
    iraf.nifs(Stdin="tmp.txt")
    iraf.gnirs(Stdin="tmp.txt")
    iraf.gemtools(Stdin="tmp.txt")
```

Or pass the `-s` flag to pytest as before:

- `pytest -s <test_file>`

Maybe I'll mark this as a new issue to fully fix at some point.

andamian reviewed (approved):
It looks good to me. Without being familiar with the pipeline, it seems to me that it achieves that it set to achieve.

Nat1405 committed:
Removes redundant os.getcwd() calls.

Nat1405 committed:
Reduces number of times a file is opened.

Nat1405 committed:
Removes assert statements and replaces with conditionals.

Nat1405 committed:
Fixes bug in 341af1cc that forgot flatdarks.

Nat1405 committed:
Refactors to attempt to remove travis_wait from .travis.yml.

In reference to Adrian's feedback on #30. Cleans up some code
that was duplicated in several tests, while I was trying to run
pytests with just 'pytest' rather than 'travis_wait XX pytest ...'

Nat1405 committed:
Sets up integration tests for new PROVENANCE extension.

This is in reference to #32.

Nat1405 commented:
Hi Sharon, thanks so much for your feedback as well. I've addressed some of your comments in 0f9c8ed, f6c8447, and 4a5d529. Much appreciated.

Nat1405 committed:
Fixes missing modules in travis ci build.

Nat1405 committed:
Adds provanance extension for calibrations.

Nat1405 commented:
Hi @dbohlender, can you take a look at my three latest commits (f3f9e6c, b10d4ef, f6c8729) and let me know what you think?

f3f9e6c adds integration tests for the new provenance extension based on NIFS program GN-2014A-Q-85. 
b10d4ef doesn't really do much, it just fixes an integration test problem.
f6c8729 adds the code that actually adds the provenance extension. Namely, it modifies the data label if there is more than one member, and creates and adds the "PROVENANCE" BINTABLE extension that contains three columns: filename, description, and type. 

Let me know if you'd like a sample of what the final tagged products look like as well. Thank you!

Nat1405 committed:
Makes CalibrationTagger variables static for ProductTagger.

Nat1405 committed:
Propagates darks through the pipeline.

Nat1405 committed:
Begins tagging of processed dark.

Nat1405 committed:
Fixes bpm not found caused by variable prefix.

Nat1405 committed:
Updates OBSTYPE of processed ronchis to be RONCHI

Nat1405 committed:
Adds integration test for full reduction.

Nat1405 committed:
Adds actual product tagger.

Nat1405 committed:
Adds integration test that should catch full reduction problems.

Nat1405 committed:
Works in Davids comments, and fixes int tests.

Nat1405 commented:
Hi @dbohlender , thanks for your review so far. I've added some new code and would love your review.

- 090bde1 makes attributes of the `CalibrationTagger` class "static" so they can be accessed by the `ProductTagger` class. This is to avoid duplicating redundant code in both of the `Tagger` classes.
- 50e723f propagates darks through the pipeline. Surprisingly easy due to good design :) 
- 5a55f63 fixes a few bugs, and is where we start using the processed dark rather than the raw darks in the provenance propagation.
- d1bf41d fixes a bug.
- 966fa9b works in your `RONCHI` comments.
- a5e4edc adds an integration test for product tagging, including **making sure the sky frames are identified properly.**
- ec95e64 changes the product tagger to add provenance information in an extension, rather than in a header.
- 16ab2cf Adds another integration test, and
- d7f1f6f incorporates your `.upper()` and readability (underscores) comments, and fixes up the integration tests.

Thanks again for your review.

Nat1405 committed:
Changes PROVENANCE extension column name to Type.

Nat1405 committed:
Adds WATN_000 removal to the taggers.

One problem with this is the WATN_000 removal happens before
the addition of the bad pixel mask (to minimize the number of times
the file is opened). However, as long as the format
of those files doesn't change, this should be okay.

Nat1405 committed:
Fixes logging in ProductTagger.

Nat1405 committed:
Fixes obstype in tagged ronchis and darks.

Nat1405 committed:
Fixes problem with undefined variable.

Nat1405 committed:
Adds previews, and fixes ProductTagger

Nat1405 committed:
Fixes failing downloads.

Removes extra quotation marks in filename from varying content-disposition headers. Fixes #35.

Nat1405 committed:
Tries to alter configuration to help programs pass for now.

Nat1405 committed:
Fixes IndexError resulting from insufficient calibrations found.

Nat1405 committed:
Makes integration tests only test uncorrected datacube production, for now.

Nat1405 committed:
Fixes gemcombine crashes with too many files.

Nat1405 committed:
Fixes typo in plus/minus one day calibration finder.

Nat1405 committed:
Changes pipeline diagnostics to reflect individual uncorrected cubes.

Nat1405 committed:
Updates diagnostics for new skyFrame identification.

Nat1405 committed:
Fixes skyframe issues.

Switches skyframe identification to use a clustering method by default. Also, removes requirement for sky subtraction for science frames. If no sky frames are found, the reduction will no longer be abandoned.

Nat1405 committed:
Removes arcdark requirement.

Removes rigid requirement for darks for arc frames. Fixes #39.

Nat1405 committed:
Improves plus/minus one day calibration checks.

Now when new calibrations are brought in from a day ahead and behind,
    they're also checked to all have the same filter, grating and
    exposure time. Fixes #37 for now.

Nat1405 committed:
Fixes some IndexErrors.

Nat1405 committed:
Improves error handling for download errors.

For #41.

Nat1405 committed:
Catches wavelength calibration errors properly.

More progress on #41.

Nat1405 committed:
Handles interactive wavelength errors.

For #41.

Nat1405 committed:
Addresses OSError from failing to copy a flat.

For #41.

Nat1405 committed:
Handles IRAF errors.

For #41.

Nat1405 committed:
Deals with too many frame problems.

For #41.

Nat1405 committed:
Fixes ValueError due to malformed EXPTIME keyword.

For #41.

Nat1405 committed:
Adds various bugfixes for 12 programs.

Nat1405 committed:
Catches more calibration reduction problems.

Nat1405 committed:
Fixes broken product tagger (skyFrameList problem).

Nat1405 committed:
Starts adding SOFTWARE, SOFT_VER, etc keywords.

Nat1405 committed:
Adds SOFTWARE, SOFT_DOI, etc to calibrations.
